docs/csharp/language-reference/operators/null-forgiving.md

```diff
@@ -0,0 +1,48 @@
+---
+title: "! (null-forgiving) operator - C# reference"
+ms.date: 10/09/2019
+f1_keywords: 
+  - "!_CSharpKeyword"
+helpviewer_keywords: 
+  - "null-forgiving operator [C#]"
+  - "! operator [C#]"
+---
+# ! (null-forgiving) operator (C# reference)
+
+Available in C# 8.0 and later, the unary postfix `!` operator is the null-forgiving operator. In an enabled [nullable annotation context](../../nullable-references.md#nullable-annotation-context), you use the null-forgiving operator to declare that expression `x` of a reference type isn't null: `x!`. The unary prefix `!` operator is a [logical negation operator](boolean-logical-operators.md#logical-negation-operator-).
+
+The null-forgiving operator has no effect at run time. It only affects the compiler's static flow analysis by changing the null state of the expression. At run time, expression `x!` evaluates to the result of the underlying expression `x`.
+
+One of the use cases of the null-forgiving operator is in testing the argument validation logic. For example, consider the following class:
+
+```csharp
+#nullable enable
+using System;
+
+public class Person
+{
+    public Person(string name) => Name = name ?? throw new ArgumentNullException(nameof(name));
+
+    public string Name { get; }
+}
+```
+
+Using the [MSTest test framework](../../../core/testing/unit-testing-with-mstest.md), you can create the following test for the constructor argument validation logic:
+
+```csharp
+[TestMethod, ExpectedException(typeof(ArgumentNullException))]
+public void NullNameShouldThrowTest()
+{
+    var person = new Person(null!);
+}
+```
```

Comment:
@BillWagner what do you think of the example? Is that good enough? Is it useful? Is there a better example? I like the current example because (1) it implicitly shows that one needs to null-check arguments even if they are of a non-nullable reference type, (2) "blows the mind" with the `null!` construct (not sure if that is a good fit for the docs, though).

Reply:
I like this sample. It demonstrates a useful practice to validate arguments at runtime even in a nullable enabled context. That reinforces the text earlier that the `!` has no runtime effect, but influences the static analysis.  And, I expect we'll see `null!` in codebases for these kinds of examples. (I've used it to annotate members that are set from deserialization as well:  `SomeType p = nul!;`

The other use case I've seen is when a type may have a custom method that checks for null, and it hasn't been annotated using `NotNullWhen(bool)`. Something like:

```csharp
Customer c = FindCustomer(id);
// c might be null here.
if (isValid(c)) // checks that c isn't null
    PrintOrders(c!.Orders);
```

Reply:
@BillWagner I think another use case is also worth to be documented. I'll update the article accordingly. Thanks!